SI-7853 Regression in explicit outer

retronym opened the pull request:
The attempt to placate pos/t4970.scala in 55c6fd4 roused another
dragon.

We've got two levers here: the type of the symbol of the outer
accessor, and the type of its DefDef.

They have been historically out of sync due to the vaguaries of
finalResultType (which is far less vague since 671e6e03c7), but
the finicky operation of ExplicitOuter now has a hard time when
we try to bring them into line.

This stuff is notoriously difficult to understand because the
trees you see from `-Xprint` show a tpt derived from the method
symbol's info, and discards the actual tpt in the tree.

Rather than letting `DefDef(acc)` call `TypeTree(sym)` and use
`sym.tpe_*.finalResultType`, this commit computes the member type
of the accessor from the current class and explicitly uses that as
the return type of the outer accessor def.

We should try to push this a little deeper. I tried to put it into
`def TypeTree`, but that broke, among others,
run/concurrent-stream.scala. Maybe `def DefDef` and `def ValDef`?

But a localised fix is the right start as it addresses the regress
in a minimal fashion to get the IDE building again.
# 

A less ad-hoc place to call memberType

Rather than localizing the fix to the outerAccessor, this
commit pushed the call to `memberType` into _all_ usages of
`newValDef` and `newDefDef`.

retronym committed:
SI-7853 An unsatisfying fix regression in explicit outer

The attempt to placate pos/t4970.scala in 55c6fd4 roused another
dragon.

We've got two levers here: the type of the symbol of the outer
accessor, and the type of its DefDef.

They have been historically out of sync due to the vaguaries of
finalResultType (which is far less vague since 671e6e03c7), but
the finicky operation of ExplicitOuter now has a hard time when
we try to bring them into line.

This stuff is notoriously difficult to understand because the
trees you see from `-Xprint` show a tpt derived from the method
symbol's info, and discards the actual tpt in the tree.

Rather than letting `DefDef(acc)` call `TypeTree(sym)` and use
`sym.tpe_*.finalResultType`, this commit computes the member type
of the accessor from the current class and explicitly uses that as
the return type of the outer accessor def.

We should try to push this a little deeper. I tried to put it into
`def TypeTree`, but that broke, among others,
run/concurrent-stream.scala. Maybe `def DefDef` and `def ValDef`?

But a localised fix is the right start as it addresses the regression
in a minimal fashion to get the IDE building again.

retronym commented:
This was a bit of a rushed push as the train was rolling into the station. I'm going to take a second   look tomorrow, so please leave this in queue

dragos commented:
..it does bring the IDE [back to life](https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala-integrate-ide/2174/)! Thanks!

retronym commented on src/compiler/scala/tools/nsc/transform/ExplicitOuter.scala:
```diff
@@ -361,9 +361,15 @@ abstract class ExplicitOuter extends InfoTransform
+        // Duelling tests pos/t7853.scala and pos/t4970.scala are somehow reconciled with this
+        // formulation. We would really like to say `DefDef(acc, Select(This(currentClass), outerField(currentClass)))`
+        val resTpe = acc.tpe.finalResultType.underlying
+        newDefDef(acc, Select(This(currentClass), outerField(currentClass)))(tpt = TypeTree(resTpe))
```
Oh, I think the real fix for this is might be to sprinkle a `mkAttributedStableRef` in here. My first attempt failed, but I'll knock my head against it some more tomorrow.

paulp commented:
@retronym 62e3a2ad82 is how I'd expect to do this. I didn't test anything beyond the identified dueling tests, but those at least don't crash.

retronym committed:
SI-7853 A less ad-hoc place to call memberType

Rather than localizing the fix to the outerAccessor, this
commit pushed the call to `memberType` into *all* usages of
`newValDef` and `newDefDef`.

The TPT of `applyOrElse` in synthetized partial functions must
be set explicitly to pass the pos/t7853-partial-function.scala.
Otherwise, the as-seen-from ends up cloning the type parameter
`B1` of `applyOrElse` as it transforms (questionably)
its bound from `List[Int @unchecked]` to `List[Int]`.

Partial Function synthesis was already a delicate area, and this
makes things more explicit which could be counted as an improvement.

retronym commented:
@paulp See any problems with the last commit? I'd like to merge this one to get the IDE build back up.

paulp commented:
LGTM.

paulp commented on src/reflect/scala/reflect/internal/Trees.scala:
```diff
@@ -589,7 +589,11 @@ trait Trees extends api.Trees { self: SymbolTable =>
-  def TypeTree(sym: Symbol): TypeTree = atPos(sym.pos.focus)(TypeTree(sym.tpe_*.finalResultType))
+  private def TypeTreeMemberType(sym: Symbol): TypeTree = {
+    // Needed for pos/t4970*.scala. See SI-7853
+    val resType = (sym.owner.thisType memberType sym).finalResultType
```
But FYI this is set up for trouble and you should probably revisit it. Taking the memberType of the symbol owner means it will be wrong whenever the symbol is inherited. When the type signature is "Symbol => TypeTree", the right thing cannot be done.

retronym replied:
Yeah, I made it private to and renamed it to limit what could get here. I can't think of any valid need to `newDefDef` or `newValDef` with an inherited symbol.

retronym replied:
You might want to rebase your Parser refactor now, then we'll can see if it passes the [ide validation build](https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala-integrate-ide/)
